From bfee4848f3dfa183a3fce849dfe77a7f12ab1d24 Mon Sep 17 00:00:00 2001
From: Michel Lind <salimma@fedoraproject.org>
Date: Fri, 7 Nov 2025 12:13:54 +0000
Subject: [PATCH] Only define `once_flag`/`call_once` if `glibc` has not
 defined them

`glibc` will define `once_flag` in the forthcoming 2.43 release, see:

https://sourceware.org/git/?p=glibc.git;a=commit;h=a7ddbf456d97ac8d1aa7afd735e196a1488bd874

Do not redefine it and related definitions if they already exist.

Fixes: #33

Signed-off-by: Michel Lind <salimma@fedoraproject.org>
---
 include/the_Foundation/c11threads.h | 2 ++
 src/c11threads.c                    | 4 ++++
 2 files changed, 6 insertions(+)

diff --git a/include/the_Foundation/c11threads.h b/include/the_Foundation/c11threads.h
index 8c762f6..70cbec3 100644
--- a/include/the_Foundation/c11threads.h
+++ b/include/the_Foundation/c11threads.h
@@ -31,7 +31,9 @@ typedef pthread_t thrd_t;
 typedef pthread_mutex_t mtx_t;
 typedef pthread_cond_t cnd_t;
 typedef pthread_key_t tss_t;
+#ifndef ONCE_FLAG_INIT
 typedef pthread_once_t once_flag;
+#endif
 
 typedef int (*thrd_start_t)(void*);
 typedef void (*tss_dtor_t)(void*);
diff --git a/src/c11threads.c b/src/c11threads.c
index 14abb05..0ab9759 100644
--- a/src/c11threads.c
+++ b/src/c11threads.c
@@ -33,7 +33,9 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.</small>
 #include <sys/time.h>
 #include <time.h>
 
+#ifndef ONCE_FLAG_INIT
 #define ONCE_FLAG_INIT PTHREAD_ONCE_INIT
+#endif
 
 #ifdef C11THREADS_NO_TIMED_MUTEX
 #  define PTHREAD_MUTEX_TIMED_NP PTHREAD_MUTEX_NORMAL
@@ -182,6 +184,8 @@ int tss_set(tss_t key, void *val) {
 
 void *tss_get(tss_t key) { return pthread_getspecific(key); }
 
+#ifndef ONCE_FLAG_INIT
 void call_once(once_flag *flag, void (*func)(void)) {
     pthread_once(flag, func);
 }
+#endif
-- 
2.47.3

